src/Juvix/Core/Erasure.hs

```diff
@@ -3,27 +3,28 @@ module Juvix.Core.Erasure
   )
 where
 
-import qualified Juvix.Core.MainLang as Core
+import qualified Juvix.Core.IR.Types as Core
+import Juvix.Core.Utility
 import qualified Juvix.EAC.Types as EAC
 import Juvix.Library hiding (empty)
 import Juvix.Utility
-import Juvix.Utility.HashMap
-import Prelude ((!!))
 
-erase' ∷ Core.CTerm → (EAC.Term, EAC.TypeAssignment)
+erase' ∷ Core.Term primTy primVal → Either ErasureError (EAC.Term, EAC.TypeAssignment)
 erase' cterm =
   let (term, env) = exec (erase cterm)
-   in (term, typeAssignment env)
+   in term >>| \term →
+        (term, typeAssignment env)
 
-exec ∷ EnvErasure a → (a, Env)
-exec (EnvEra env) = runState env (Env empty 0 [])
+exec ∷ EnvErasure a → (Either ErasureError a, Env)
+exec (EnvEra env) = runState (runExceptT env) (Env empty 0 [])
```

Comment:
Good change, I suspect a lot of our effects that lack `runExceptT` will slowly gain them over time